### PR TITLE
cumulative acks

### PR DESCRIPTION
**Pull request by kazuho**
ToDo:
* [x] cap the size of `ack_queue`
* [x] send ACK either when we need to, or when there's new information
* [x] report ack_delay accurately (instead of over-reporting at the rate of 1024/1000)
* [x] ~~preserve `largest_pn_received` along with `largest_pn_received_at` and use it for building ACK. Otherwise, we will report incorrect ack_delay if ACK of `largest_pn_received` gets ACKed before a PN that is smaller (happens when there is reorder).~~

### CODE REVIEW

**Comment by kazuho** [comment]
@janaiyengar Would you be interested in reviewing the PR before I merge?

Some notable design decisions:
* Max ACK block count is capped to 62 (instead of 100). This simplifies the ACK frame encoder because block count below 64 can be expressed in one byte. The maximum is _not_ 63 due to the following point.
* We continue ACKing the largest PN received even when we know that the peer has received the ACK for the PN, in order to avoid reporting incorrect ack_delay (see e4af6a5).
* We no longer split ACK frames into multiple packets if there is not enough space. Instead, we rollback the write to the current packet when it runs out of space, and write the entire ACK into a newly allocated packet (see 8c8a27c)

**Comment by kazuho** [comment]
> We continue ACKing the largest PN received even when we know that the peer has received the ACK for the PN, in order to avoid reporting incorrect ack_delay (see e4af6a5).

Me notices that this does not happen, because an ACK of ACK always arrive in a packet with a PN that is greater than the PN being ACK-ACKed.

File: lib/quicly.c [hunk 1]
```diff
@@ -2375,15 +2445,24 @@ static int do_detect_loss(quicly_loss_t *ld, uint64_t largest_pn, uint32_t delay
     }
 
     /* schedule early retransmit alarm if there is a packet outstanding that is smaller than largest_pn */
-    if (ack->packet_number < largest_pn && ack->sent_at != INT64_MAX && ack->is_alive)
-        *loss_time = ack->sent_at + delay_until_lost;
+    while (ack->packet_number < largest_pn && ack->sent_at != INT64_MAX) {
```
Comment (janaiyengar) [inline lib/quicly.c]: This just changed in #1974, but we can do this in a separate PR.

File: lib/quicly.c [hunk 2]
```diff
@@ -2918,13 +2991,16 @@ static int handle_ack_frame(quicly_conn_t *conn, size_t epoch, quicly_ack_frame_
...
     /* OnPacketAckedCC */
     uint32_t cc_type = 0;
-    /* TODO (jri): this function should be called for every packet newly acked. */
-    if (quicly_loss_on_packet_acked(&conn->egress.loss, frame->largest_acknowledged)) {
-        cc_type = CC_RTO;
-        conn->egress.cc.in_first_rto = 0;
-    } else if (conn->egress.cc.in_first_rto) {
-        cc_type = CC_RTO_ERR;
-        conn->egress.cc.in_first_rto = 0;
+    /* TODO (jri): this function should be called for every packet newly acked. (kazuho) I do not think so;
```
Comment (janaiyengar) [inline lib/quicly.c]: We should probably discuss this. I thought this was OnPacketAcked.

File: lib/quicly.c [hunk 3]
```diff
@@ -2918,13 +2991,16 @@ static int handle_ack_frame(quicly_conn_t *conn, size_t epoch, quicly_ack_frame_
...
+     * quicly_loss_on_packet_acked is NOT OnPacketAcked */
+    if (smallest_newly_acked != UINT64_MAX) {
```
Comment (janaiyengar) [inline lib/quicly.c]: We may want to change this behavior as well... I think we're probably moving away from looking for newly acked packets in the ACK frame... This is also a change in PR 2040 and 1974.

**Comment by kazuho** [comment]
Thank you for checking. As you say, let's merge the PR and address the remaining issues separately.
